### PR TITLE
Record basic performance timing events with graphite

### PR DESCRIPTION
**Pull request by Julesssss**
CC @tgolen 

Adds a simple event Timing functionality which shares results to Grafana with our Graphite API. Three events are tracked: `homepage_initial_render`, `homepage_reports_loaded`, `switch_report.pinned`, `switch_report.unpinned` 

- We are using `Platform.OS` as the referrer, which will be handled by the Web API
- The events will not yet be persisted to Graphite, as we must first remove the [Web API whitelist here](https://github.com/Expensify/Web-Expensify/blob/21e617aae3766533fc8e17830cb8ff3073636a3a/api.php#L2030).  

### Fixed Issues
$ Fixes https://github.com/Expensify/Expensify/issues/147003

### Tests

**Setup**
Apply the following diff: 
```
diff --git a/src/libs/Timing.js b/src/libs/Timing.js
index 5ef17b21..b83f30f1 100644
--- a/src/libs/Timing.js
+++ b/src/libs/Timing.js
@@ -8,6 +8,7 @@ let timestampData = {};
  * @param {String} eventName
  */
 function start(eventName) {
+    console.debug(`Timing.start(${eventName})`);
     timestampData[eventName] = Date.now();
 }
 
@@ -19,6 +20,7 @@ function start(eventName) {
 function end(eventName) {
     if (eventName in timestampData) {
         const eventTime = Date.now() - timestampData[eventName];
+        console.debug(`Timing.end(${eventName}), time: ${eventTime}`);
 
         Graphite_Timer({
             name: eventName,

```

**A) Ensure that homepage events are timed correctly**
1. Build to each platform, sign out
2. Filter console output with 'Timing'
3. Sign In
4. The following events should be timed: `expensify.cash.homepage_initial_render`, `expensify.cash.homepage_reports_loaded`
5. Refresh the page
6. Confirm the events are timed once more 

**B) Ensure that switch report event is timed correctly**
1. Select a previously pinned chat that is not currently selected
2. The following event should be timed: `expensify.cash.switch_report.pinned`
3. Using the ChatSwitcher, start or find a group chat with multiple participants that is not currently active/pinned
4. The following event should be timed: `expensify.cash.switch_report.unpinned`
5. Using the chat switcher, search for a user who is **not currently active/pinned**
6. The following event should be timed: `expensify.cash.switch_report.unpinned`
7. Using the chat switcher, search for a user who is currently **pinned**
8. The following event should be timed: `expensify.cash.switch_report.pinned`

**C) Ensure that Graphite API request is made**
1. With a network inspector, confirm that requests are made to 'api?command=Graphite_Timer', for each of the above events, with the following fields: 
    ```name, value, referer```
2. Referer should match Web, iOS or Android _-- Note: Desktop is also reported as Web_

Screenshot 2020-12-17 at 20 20 18
Screenshot 2020-12-17 at 20 20 34
Screenshot 2020-12-17 at 20 22 53

### Screenshots
N/A

### CODE REVIEW

**Review by tgolen — CHANGES_REQUESTED** [review]
Does this work on iOS and Android? I'm not sure that `window` exists at all, or if it might crash those platforms. 

I think it would be better to wrap all the performance stuff up into its own action file. This will:
- Remove a lot of global references to `window` that are scattered around
- DRY up the code a bit
- Make it easier to remove at some point (I am assuming this is not permanent)
- Simplify the code

**Review by chiragsalian — CHANGES_REQUESTED** [review]
I don't think your PR is working well when I tested it. So when I first open the app and see the signing form I notice that time has started recording,
image

Then I did this review and signed in and the end times were massive,
image

That's 654 seconds which is 10mins. This can heavily skew our grafana data and hence we should begin recording after signing in.

Additionally while switching between chats i don't see anything recorded when I switch to an existing pinned chat in LHN. If I switch to a pinned chat using chat switcher then I see only timing.start with no timing.end recorded
image

The only time I see timing.start and end recorded for `switch_report` is when I use chat switcher to select an unpinned chat so you would have to retest and update the locations you make these calls. (I see now in the test you mentioned to search for a chat that's not currently in the LHN, why not record every time chat is switched)

Always try to number your tests. Makes it easier to say update point 5 than update,
> Filter console output with 'recordTimingEvent'

also wrt to that test I think you meant filter output with `Timing.` because I don't see what `recordTimingEvent` is.

**Review by tgolen — CHANGES_REQUESTED** [review]

**Review by tgolen — CHANGES_REQUESTED** [review]
I tested this and it looks like it works really well. I initially had the "report loaded" event fire really late (it took nearly 43 seconds) but I was never able to reproduce that.

As far as having two different events for switching reports, I think you're onto something. I saw that a pinned report takes a lot less time to switch than a non-pinned report, and I think it's good to separate that data. What I would suggest is just adding onto the end of the event with `.pinned` and `.unpinned`. That way, in Grafana you can always combine them with `.*` if we want and it's very flexible.

**Review by chiragsalian — DISMISSED** [review]
Code LGTM and works well so I'm gonna approve but yeah like Tim said I would really like this to be captured,
```
I saw that a pinned report takes a lot less time to switch than a non-pinned report, and I think it's good to separate that data. What I would suggest is just adding onto the end of the event with .pinned and .unpinned
```

**Review by tgolen — CHANGES_REQUESTED** [review]

File: src/pages/home/report/ReportView.js [hunk 1]
```diff
@@ -20,6 +22,14 @@ const propTypes = {
 class ReportView extends React.PureComponent {
     componentDidMount() {
         subscribeToReportTypingEvents(this.props.reportID);
+
+        Timing.end(CONST.TIMING.SWITCH_REPORT, CONST.TIMING.UNPINNED);
+    }
+
+    componentDidUpdate(props) {
+        if (!props.isActiveReport) {
+            Timing.end(CONST.TIMING.SWITCH_REPORT, CONST.TIMING.PINNED);
```
[the remark this answers is not in the document]
Comment (tgolen) [reply src/pages/home/report/ReportView.js:31]: Ah, OK. Makes sense for now. I think another solution that might be more scalable at some point is to have the params to the `end()` method be `nameOfEventToEnd, nameOfEventToSendToGrafana`. Total NAB for now though.

**Review by tgolen — APPROVED** [review]

**Pull request merged**